### ISSUE TITLE
Watch errors when Getting an object that does not implement the Watch verb

### ISSUE THREAD

**Issue by miguelsorianod** [operator-framework/operator-sdk#2649]
Hi,

We have added some code in our operator that manages ImageStreamTag objects (an OpenShift Kubernetes object).

The code performs Gets of ImageStreamTag objects and deletes them.

We execute the operator with operator-sdk run --local and the code seems to Get the objects and delete them in a correct way but we observe the following errors are being logged in the operator constantly:
```
E0313 12:27:07.868299    9878 reflector.go:280] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to watch *v1.ImageStreamTag: the server does not allow this method on the requested resource (get imagestreamtags.image.openshift.io)
E0313 12:27:20.990531    9878 reflector.go:280] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to watch *v1.ImageStreamTag: the server does not allow this method on the requested resource (get imagestreamtags.image.openshift.io)
E0313 12:27:21.997340    9878 reflector.go:280] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to watch *v1.ImageStreamTag: the server does not allow this method on the requested resource (get imagestreamtags.image.openshift.io)
E0313 12:27:23.008849    9878 reflector.go:280] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to watch *v1.ImageStreamTag: the server does not allow this method on the requested resource (get imagestreamtags.image.openshift.io)
```

With the api-resources subcommand we have found that ImageStreamTag objects do not implement the 'watch' verb by themselves as you can observe in the following command output:
```
msoriano@localhost:~/go/src/github.com/3scale/3scale-operator (change-imagestream-tagging-structure)$ oc api-resources -o wide | grep ImageStream
imagestreamimages                     isimage        image.openshift.io                      true         ImageStreamImage                     [get]
imagestreamimports                                   image.openshift.io                      true         ImageStreamImport                    [create]
imagestreammappings                                  image.openshift.io                      true         ImageStreamMapping                   [create]
imagestreams                          is             image.openshift.io                      true         ImageStream                          [create delete deletecollection get list patch update watch]
imagestreamtags                       istag          image.openshift.io                      true         ImageStreamTag                       [create delete get list patch update]
```

It seems as if the operator-sdk client tries to always add Watches to the objects where a Get is performed. When the object being Getted does not implement the watch verb this causes the observed logged errors in the operator.

The operator-sdk version used is:
```
msoriano@localhost:~/go/src/github.com/miguelsorianod$ operator-sdk version
operator-sdk version: "v0.15.2", commit: "ffaf278993c8fcb00c6f527c9f20091eb8dd3352", go version: "go1.13.3 linux/amd64"
```

**Comment by camilamacedo86** [comment]
Hi @miguelsorianod, 

If you are adding third-party schemas (OCP API), then you need to customize the metrics default implementation. See the [doc](https://github.com/camilamacedo86/operator-sdk/blob/master/doc/user-guide.md#default-metrics-exported-with-3rd-party-resource).

Similar to issues of #2577, #1858. 

I am closing this one as a duplicate. However, feel free to ping us to re-open if you think that is required.

**Comment by miguelsorianod** [comment]
Hi @camilamacedo86 ,

I think this issue is not related or duplicated of one of the ones you linked.
Those are related to metrics and we are not experiencing that problem in the metrics implementation.

It seems related to objects (being in OpenShift or Kubernetes objects themselves) that do not implement the Watch verb and the client being used.

**Comment by miguelsorianod** [comment]
To try to add more information. A way to reproduce this would be:
1. Try to perform a Get of an ImageStreamTag object using the client provided by the operator-sdk. Instead of an ImageStreamTag, any object that does not implement the 'watch' verb would reproduce the problem. You can check what verbs implement each object type by executing kubectl api-resources -o wide and looking at the list of verbs that each object has
1. Observe how in the operator logs the errors we showed in the initial comment but with the type of object that has been getted

**Comment by camilamacedo86** [comment]
Hi @miguelsorianod, 

You have been facing the error: `Failed to watch *v1.ImageStreamTag`
And when you checked this OCP api it is not implementing the watch verb. 

`imagestreamtags                       istag          image.openshift.io                      true         ImageStreamTag                       [create delete get list patch update]
`

1. Could you please let us know if this issue occurs if you just startup the operator whiteout apply any CR used by it? I mean, before any reconciliation be performed. 

2. Also, could you let us know if you customize the `filteredGVK, err := k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)` by removing this new API as well your main.go? If not, could you please do that and let us know if the issue will still occurs?

**Comment by miguelsorianod** [comment]
Hi,

The issue does not occur when I start the operator, without any CR being deployed.

The second point shouldn't have anything to do, as the metrics configuration/initialization endpoint happen at startup time (in cmd/manager/main.go). If the error was to be related to metrics and the filtering of GKVs it would be directly shown at operator startup time.

Related to the third point, we have not implemented any explicit watch of the ImageStreamTag type.

**Comment by camilamacedo86** [comment]
Hi @miguelsorianod, 

Really thank you for your information provided. In this case, we need to do a POC to isolate the scenario first and ensure that the same error would be faced with the default scaffold. Which means:

- On top of the [Memcached Go sample](https://github.com/operator-framework/operator-sdk-samples/tree/master/go/memcached-operator) and the schema ImageStreamTag
- Then, customize the `filteredGVK, err := k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)` to remove it from it. 

Run the operator and check if the same issue will be faced whiteout any additional change. If yes, then we can confirm that the client go provided by SDK/controller-runtime is looking for the watch verb of all API's added on it. Then, we will need to figure out where and why it has been made to know how better address it. May it requires a change in the controller-runtime and/or client-go instead of SDK. 

Could you please help us here by providing this POC?

**Comment by miguelsorianod** [comment]
Hi @camilamacedo86,

I've updated the first comment with the used operator-sdk version.

I've created a POC in https://github.com/miguelsorianod/pocs-operator-sdk/commits/watch-issue-poc
that contains a new operator-sdk project (v0.15.2) with an example API version `poc.example.com/v1alpha1` and a kind `PocKind` and its respective controller.

The code:
* Makes sure the filtered GKV list is an empty array of GKV to avoid any possible metrics registering/permissions problem (see cmd/manager/main.go)
* Registers the image.openshift.io APIGroup schemas to be able to use ImageStreamTags (see cmd/manager/main.go)
* In the controller code we try to get an ImageStreamTag. When the get is performed an error not found is returned but that is expected in this POC example as the object does not really exist but it is enough to reproduce the problem.

If you run the operator with `operator-sdk run --local` (In my case I've run it as kube:admin user that has the `admin` ClusterRole associated) and then create an example `PocKind` CR the reported errors in the initial commit should be shown.

**Comment by camilamacedo86** [comment]
Really thank you for the collab and for providing it. I will be working on it as soon as possible for we reach a solution.

**Comment by camilamacedo86** [comment]
Hi @miguelsorianod, 

I could check that just add the schema that has not the watch verb implemented to the manager as follows will cause this scenario. 

**just add the schema without the watch verb will cause this issue**

Shows that when we add the Schema to the manager, the client/cache use/provided will require the `watch` verb. 

```go
import (
...
         imagev1 "github.com/openshift/api/image/v1"
...
)

func main() {
         ....
	// Setup Scheme for OpenShift imagestreams and related
	if err := imagev1.AddToScheme(mgr.GetScheme()); err != nil {
		log.Error(err, "")
		os.Exit(1)
	}
 
```

**Possible Solution with client.Reader**

So, I understood that to use the `client.Reader` instead of would be a possible as follows.

```go
import (
...
         imagev1 "github.com/openshift/api/image/v1"
...
)

// newReconciler returns a new reconcile.Reconciler
func newReconciler(mgr manager.Manager) reconcile.Reconciler {
	return &ReconcilePocKind{client: mgr.GetClient(), scheme: mgr.GetScheme(), APIReader: mgr.GetAPIReader()}
}

...

// ReconcilePocKind reconciles a PocKind object
type ReconcilePocKind struct {
	// This client, initialized using mgr.Client() above, is a split client
	// that reads objects from the cache and writes to the apiserver
	client client.Client
	APIReader client.Reader // the client that will not use the cache
	scheme *runtime.Scheme
}
 

func (r *ReconcilePocKind) Reconcile(request reconcile.Request) (reconcile.Result, error) {
	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
	reqLogger.Info("Reconciling PocKind")

	....

	// Get the latest resource to prevent updating a stale status.
	imageStreamTag := &imagev1.ImageStreamTag{}
	imageStreamTagName := "example-istag"
	namespacedName := types.NamespacedName{Namespace: "request.NamespacedName", Name: imageStreamTagName}
	if err := r.APIReader.Get(context.TODO(), namespacedName, imageStreamTag); err != nil {
		reqLogger.Error(err, "unexpected error")
	}

	return reconcile.Result{}, nil
}
```

**Problem faced with the possible solution**

However, I am facing the following error when I apply the CR. @joelanford, @fabianvf Am I missing something here? 

> "error":"no kind is registered for the type v1.ImageStreamTag in scheme \"pkg/runtime/scheme.go:101\""


**What is missing to address here?**

- [ ] Figure out what if is missing something in the ` client.Reader` solution in order to make it works and/or the confirmation from @joelanford @fabianvf that their suggestion will not attend this scenario.  
- [ ] check if we cannot address improvements regards this scenario in the controller-runtime/go-client
- [ ] address its result/solution in the docs
- [ ] raise an issue to check if OCP cannot implement the watch for this api. See: https://github.com/openshift/api/issues/601

**Comment by camilamacedo86** [comment]
Hi @miguelsorianod, 

I could check that the client cache will require the List and Watch permissions. So, the suggested solution would be using the APIReader [client.Reader](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#Reader) instead of. 

See my poc in : https://github.com/camilamacedo86/operator-sdk-samples/tree/poc-watch/go/memcached-operator 

Just clone the branch and then run `make install`. The namespace `memcached` will be created and all will be installed there. 

And then, see that the error faced is because the object is not found:

> ,"msg":"unexpected error","Request.Namespace":"memcached","Request.Name":"example-memcached","error":"ImageStreamTags must be retrieved with <name>:<tag>"

Note that, the ImageStreamTags API in this scenario shows not be returning `HTTP Status code 404` and then, because of this  the log is not done in the [IsNotFound](https://github.com/camilamacedo86/operator-sdk-samples/blob/poc-watch/go/memcached-operator/pkg/controller/memcached/memcached_controller.go#L118)

```
// IsNotFound returns true if the specified error was created by NewNotFound.
func IsNotFound(err error) bool {
	return ReasonForError(err) == metav1.StatusReasonNotFound
}
```

**So, IMO: what was done and missing to address here?**

- [x] Create a POC with` client.Reader's` as a solution to attend this scenario.  See:  https://github.com/camilamacedo86/operator-sdk-samples/tree/poc-watch/go/memcached-operator 
- [x] Raise an issue to check if OCP cannot implement the watch for its API. See: https://github.com/openshift/api/issues/601
- [ ] address its result/solution in the docs

Please, let me know if it attends your need and answers your question/scenario.

**Comment by miguelsorianod** [comment]
Hi @camilamacedo86,

> I could check that the client cache will require the List and Watch permissions. So, the suggested solution would be using the APIReader client.Reader instead of.

Yes. We have an implementation where we workaround the cache at the moment for our use case: https://github.com/3scale/3scale-operator/blob/bab0608b9e36953e0eb0259ad348fd1b91e99f68/pkg/3scale/amp/operator/upgrade.go#L562

It would be great if we could take advantage of the use of the cache instead of having to provide the non-cached client reader. I've seen you opened an issue in the OpenShift project https://github.com/openshift/api/issues/601 related to this.

I don't know how frequent this use case is but it might be a good idea to make aware users of operator-sdk about this at the moment.

**Comment by camilamacedo86** [comment]
Just to clarifies.The issue faced is:

```
{"level":"info","ts":1584718937.6659641,"logger":"controller_memcached","msg":"Reconciling Memcached.","Request.Namespace":"memcached","Request.Name":"example-memcached"}
E0320 15:42:17.676888       1 reflector.go:280] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to watch *v1.ImageStreamTag: unknown (get imagestreamtags.image.openshift.io)
{"level":"info","ts":1584718937.766342,"logger":"controller_memcached","msg":"ImageStreamTag resource not found. Ignoring since object must be deleted.","Request.Namespace":"memcached","Request.Name":"example-memcached"}
{"level":"error","ts":1584718937.766376,"logger":"controller_memcached","msg":"Failed to get ImageStreamTag.","Request.Namespace":"memcached","Request.Name":"example-memcached","error":"ImageStreamTag.image.openshift.io \"example-imagetag\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/camilamacedo/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/operator-framework/operator-sdk-samples/go/memcached-operator/pkg/controller/memcached.(*ReconcileMemcached).Reconcile\n\tmemcached-operator/pkg/controller/memcached/memcached_controller.go:135\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/camilamacedo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/camilamacedo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/camilamacedo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/camilamacedo/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/camilamacedo/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/camilamacedo/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
```

**Comment by camilamacedo86** [comment]
Regards the request. For it returns NOT FOUND and then, is required to:

```
	// Get the latest resource to prevent updating a stale status.
	img := &imagev1.ImageStreamTag{}
	err = r.APIReader.Get(context.TODO(), types.NamespacedName{Name: fmt.Sprintf("%s:%s", "example-imagetag", "example-tag") , Namespace: memcached.Namespace}, img)
	if err != nil {
		if errors.IsNotFound(err) {
			reqLogger.Info("worked not found")
		} else {
			reqLogger.Error(err, "unexpected error")
		}
	}
```
Otherwise, it will not mount correct the request and the status code will be 400. Shows that OCP is not following the standards in this scenario.

**Comment by camilamacedo86** [comment]
HI @miguelsorianod, 

So, we: 
- Open an issue to see if OCP can implement the WACTH for this API. https://github.com/openshift/api/issues/601
- Open an issue to start a thread and see if we can improve the client-go/controller-runtime to handle this scenario and skip the cache: https://github.com/kubernetes-sigs/controller-runtime/issues/870
- And then, we have an OpenPR to explain what to do when it is faced. See: https://github.com/operator-framework/operator-sdk/issues/2649 

Also, we could confirm that you have been using the workaround solution so far as well. 
In this way, I do not see any further action that we could take here and this issue will be closed with the merge of #2649.

However, please feel free to ping us and let us know if you have any objection or see if has anything else that should be addressed here.
